packages/aws-cdk/lib/api/hotswap/cloudwatch-logs.ts

```diff
@@ -0,0 +1,97 @@
+  // it assumes the bootstrap deploy Role, which doesn't have permissions to update Lambda functions
+  const sdk = await sdkProvider.forEnvironment(resolvedEnv, Mode.ForWriting);
+
+  // add the sdk to the hotswapLogMonitor so that it also has the
+  // correct credentials to monitor CloudWatch log groups
+  hotswapLogMonitor.addSdk(sdk);
+  // The current resources of the Stack.
+  // We need them to figure out the physical name of a resource in case it wasn't specified by the user.
+  // We fetch it lazily, to save a service call, in case all hotswapped resources have their physical names set.
+  const listStackResources = new LazyListStackResources(sdk, stackArtifact.stackName);
+  const evaluateCfnTemplate = new EvaluateCloudFormationTemplate({
+    stackArtifact,
+    parameters: assetParams,
+    account: resolvedEnv.account,
+    region: resolvedEnv.region,
+    partition: (await sdk.currentAccount()).partition,
+    urlSuffix: sdk.getEndpointSuffix,
+    listStackResources,
+  });
+
+  const resources = await listStackResources.listStackResources();
+
+  // do a first pass at identifying all log groups
+  for (const resource of resources) {
+    if (resource.ResourceType === 'AWS::Logs::LogGroup') {
+      logGroupLogicalIds.add(resource.LogicalResourceId);
+      // track all the resources that will create a log group if one
+      // is not provided
+    } else if (implicitTypes.includes(resource.ResourceType)) {
+      const pathPart = resource.ResourceType.split('::')[1].toLowerCase();
+      implicitMap.set(resource.LogicalResourceId, pathPart);
+    }
+  }
+
+  // For each CloudWatch Log group in the template, find all resources that
+  // reference that log group and if it matches once of the included types, add it.
+  for (const id of logGroupLogicalIds) {
+    const resourcesReferencingLogGroup = evaluateCfnTemplate.findReferencesTo(id);
+    for (const r of resourcesReferencingLogGroup) {
+      if (includedTypes.includes(r.Type)) {
+        const groupName = await evaluateCfnTemplate.findPhysicalNameFor(id);
+        if (groupName) {
+          logGroupNames.push(groupName);
+        }
+        // we are explicitely creating a cloudwatch log group for this resource
+        // so add that log group and remove the resource from the list of resources
+        // that might have an implicitely created log group
+      } else if (implicitTypes.includes(r.Type)) {
+        implicitMap.delete(r.LogicalId);
+        const groupName = await evaluateCfnTemplate.findPhysicalNameFor(id);
+        if (groupName) {
+          logGroupNames.push(groupName);
+        }
+      }
+    };
+  }
+
+  // some resources can be created with a custom log group (handled above).
+  // if a custom log group is not created, then the service will create one with a
+  // specific name i.e. '/aws/codebuild/project-name'
+  for (const [k, v] of implicitMap) {
```

Comment:
Yeah, some better names than `k` and `v` would be nice here 😉. `logicalId` and `serviceName` perhaps?